File: pglite/src/pglite-custom/release/runtime/syscalls.ts
export interface SyscallRecord {
	name: string;
	args: unknown[];
	result: unknown;
}

export interface SyscallRecorder {
	record(name: string, args: unknown[], result: unknown): void;
	recent(): SyscallRecord[];
}

export function createSyscallRecorder(maxEntries = 256): SyscallRecorder {
	const entries: SyscallRecord[] = [];
	return {
		record(name: string, args: unknown[], result: unknown): void {
			entries.push({ name, args, result });
			if (entries.length > maxEntries) {
				entries.shift();
			}
		},
		recent(): SyscallRecord[] {
			return entries.slice();
		},
	};
}

File: src/main/java/io/github/hidekatsu_izuno/pglite_jdbc/pglite/release/runtime/syscalls.java
package io.github.hidekatsu_izuno.pglite_jdbc.pglite.release.runtime;

import java.util.ArrayDeque;
import java.util.List;

public class syscalls {
    public record SyscallRecord(String name, Object[] args, Object result) {}

    public interface SyscallRecorder {
        void record(String name, Object[] args, Object result);

        List<SyscallRecord> recent();
    }

    private syscalls() {}

    public static SyscallRecorder createSyscallRecorder(int maxEntries) {
        var entries = new ArrayDeque<SyscallRecord>();
        return new SyscallRecorder() {
            @Override
            public void record(String name, Object[] args, Object result) {
                entries.addLast(new SyscallRecord(name, args, result));
                while (entries.size() > maxEntries) {
                    entries.removeFirst();
                }
            }

            @Override
            public List<SyscallRecord> recent() {
                return List.copyOf(entries);
            }
        };
    }
}
